DRA kubelet: refactoring

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

- Refactoring (first commit)
- Renaming of "plugin name" to "driver name" (second commit)

The rest of the system logs information using "driverName" as key in structured logging. The kubelet should do the same. This also removes the implicit assumption that "plugin name" == "driver name" from all code (one less thing readers must remember).

#### Which issue(s) this PR fixes:

Related-to: https://github.com/kubernetes/kubernetes/pull/132058

#### Special notes for your reviewer:

This is in preparation of https://github.com/kubernetes/kubernetes/pull/132058.

#### Does this PR introduce a user-facing change?
```release-note
DRA kubelet: logging now uses `driverName` like the rest of the Kubernetes components, instead of `pluginName`.
```

/assign @klueska

pohly committed:
DRA kubelet: code cleanup

Replaced the Manager interface with a simple struct. There was no need for the
interface.

Replaced the global instance of the DRA plugins store with a fresh instanced
owned by the manager. This makes unit testing a bit easier (no need to restore
state, would enable parallelizing long-running tests).

Simplified the plugin.PluginsStore type to just "plugin.Store" because it
stuttered. The Plugin type is kept because having one struct named after its
package is one common exception from the "don't stutter" guideline.
For the sake of clarify, the package gets imported as "draplugin" (=
<parent dir>/<package>).

Removed unused NewManager "node" parameter and replaced direct construction of
cache and manager with calls to NewManager because that is how the manager
should get constructed (less code, too).

Fixed incorrect description of Manager: the plugin store is the entity which
manages drivers, not the manager. The manager is focused on the DRA logic
around ResourceClaims.

pohly committed:
DRA kubelet: move RegistrationHandler into Store

This moves wiping into the plugins store. The advantages are:
- Only a single mutex is needed in the plugin store.
- The code which decides about queuing and canceling wiping
  has access to all relevant information while that information
  is protected against concurrent modifications.
- It prepares for making that code more complex ("connection
  monitoring").

In retrospect it is not clear whether the RegistrationHandler did the right
thing in all cases when endpoints got registered and removed concurrently.

The RegistrationHandler became a thin, stateless layer on top of the store.
Because it's not really needed anymore, the required methods for the plugin
manager now get provided directly by the store.

The disadvantage is the slightly more complex initialization, but that was also
a problem before which just hadn't been solved: wiping ran without the context
of the manager. Now it does.

pohly committed:
DRA kubelet: clarify plugin vs, driver name

The rest of the system logs information using "driverName" as key in structured
logging. The kubelet should do the same.

This also gets clarified in the code, together with using consistent a
consistent name for a Plugin pointer: "plugin" instead of "client" or
"instance".

The New in NewDRAPluginClient made no sense because it's not constructing
anything, and it returns a plugin, not a client -> GetDRAPlugin.

pohly commented:
/priority important-soon

This blocks https://github.com/kubernetes/kubernetes/pull/132058, which is needed for DRA GA.

bart0sh commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -268,17 +268,17 @@ func (m *Manager) prepareResources(ctx context.Context, pod *v1.Pod) error {
 				continue
 			}
-			client, err := m.draPlugins.NewDRAPluginClient(driverName)
+			plugin, err := m.draPlugins.GetDRAPlugin(driverName)
```
Is there any documentation that explains that DRA plugin is a part of Kubelet and DRA driver is an external component to which DRA plugin is connecting?

pohly replied:
The terminology used elsewhere is that a DRA driver "provides (or deploys) a kubelet plugin".

I don't think we say that "DRA plugin is a part of Kubelet" or "DRA driver is an external component to which DRA plugin is connecting".

klueska replied:
I am fine with / actually prefer the term plugin here. The driver is the whole thing -- this is the kubelet plugin part of that.

bart0sh commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -184,29 +184,32 @@ func (h *RegistrationHandler) wipeResourceSlices(ctx context.Context, delay time
+	chosenService, err := h.validateSupportedServices(driverName, supportedServices)
 	if err != nil {
-		return fmt.Errorf("version check of plugin %s failed: %w", pluginName, err)
+		return fmt.Errorf("invalid supported gRPC versions of DRA driver plugin %s at endpoint %s: %w", driverName, endpoint, err)
```
is it version or service?

bart0sh replied:
Would it make sense to use `service` in the error message to match the variable name or rename the API to `validateSupportedVersions`?

bart0sh commented:
> The rest of the system logs information using "driverName" as key in structured logging.

Which system? CSI and Device plugins use "plugin" term quite widely:
```
$ git grep -i plugin ./pkg/kubelet/cm/devicemanager | wc -l
     606
$ git grep -vi plugin ./pkg/kubelet/volumemanager |wc -l
   11947
```
Not speaking about `pkg/kubelet/pluginmanager`.

pohly commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -23,20 +23,37 @@ import (
+//
+// It returns an informative error message including the driver name
+// with an explanation why the driver is not usable.
+func (s *Store) NewDRAPluginClient(driverName string) (*Plugin, error) {
```
This is not worse than before, where the object was [constructed by the registration handler](https://github.com/kubernetes/kubernetes/blob/8d3fb9ee0a51b6a6ea135d991391c35806422c19/pkg/kubelet/cm/dra/plugin/registration.go#L219-L227)...

I was considering putting this into plugin.go, but there are some dependencies between store and plugin for connection management which make it simpler to put this here.

bart0sh replied:
Connection management should be either part of plugin or, better, of another entity. Adding it to the store doesn't sound good to me either.
Object creation and it's functionality are two different things. I'm not sure we should compare them.

pohly commented on pkg/kubelet/cm/dra/plugin/plugin.go:
```diff
@@ -35,8 +35,11 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/metrics"
 )
 
+// Plugin contains information about one registered plugin of a DRA driver.
```
A DRA driver is a higher-level concept. It does register one or more plugins. A driver is identified by its name. A plugin is identified by that same name and the endpoint.

I don't think we we should say that a driver registers itself, because what is "it" in that case?

bart0sh replied:
That sounds logical to me, thanks for the explanations. It still contradicts with Device Plugins terminology and (may be) with the CSI one though.

klueska replied:
this explanation / comment matches my mental model and sounds good to me.

pohly commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -281,9 +284,9 @@ func (h *RegistrationHandler) validateSupportedServices(pluginName string, suppo
-func (h *RegistrationHandler) DeRegisterPlugin(pluginName, endpoint string) {
-	if p, last := h.draPlugins.remove(pluginName, endpoint); p != nil {
-		// This logger includes endpoint and pluginName.
+func (h *RegistrationHandler) DeRegisterPlugin(driverName, endpoint string) {
```
I can change this particular parameter to "pluginName" and then map it to "driverName" elsewhere in the code, with the explanation that DRA drivers must register themselves with their driver name.

Would that be better?

bart0sh replied:
I'm not sure it will be better. After your explanation above we need to decide if we want to stick with the concept of "driver registers one or more plugins" or we want to keep using the terms that were used before for other types of plugins.

pohly replied:
Registering multiple endpoints under the same plugin name (using terminology of the registration API here) is a new thing. We cannot rely on other types of plugins for guidance and have to come up with some consistent terminology ourselves.

bart0sh replied:
That makes sense to me now, thanks for the explanations!

klueska replied:
:+1

pohly commented:
> > The rest of the system logs information using "driverName" as key in structured logging.
> 
> Which system? CSI and Device plugins use "plugin" term quite widely:

Of course the term is used. The question is what it stands for and whether it makes sense here for DRA.

All other DRA components and APIs use "driver name". Even the kubelet metric uses "driver name" as label.

CSI also uses "driver name". They had a separate "plugin name" parameter for CSI registrar and discarded it because it made no sense to use anything other than the "driver name".

pohly commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -23,20 +23,37 @@ import (
+	if driverName == "" {
+		return nil, errors.New("DRA driver name is empty")
+	}
+	client := s.get(driverName)
```
No. "null" != "nil".

bart0sh commented on pkg/kubelet/cm/dra/types.go:
```diff
@@ -17,43 +17,9 @@ limitations under the License.
-
-	// GetContainerClaimInfos gets Container ClaimInfo objects
-	GetContainerClaimInfos(pod *v1.Pod, container *v1.Container) ([]*ClaimInfo, error)
-}
```
This was added to simplify mocking in the unit tests, but seems to not used. I agree, it can be removed, unless we come up with unit tests that want to mock Manager, e.g. if they don't want to run gRPC server.

klueska replied:
+1

bart0sh commented:
@pohly 
> CSI also uses "driver name".

Is this term is the same for CSI and DRA? Technically all 3 types of kubelet plugins (or drivers?) are similar - they can register more than one plugin with the same name, but diffferent sockets, they have common registration mechaninsm etc. 
If CSI and Device Plugins use different terms (plugin and driver) and DRA wants to use either of them it's ok. If DRA wants to use something different, I'm not sure we want to add to the mess we already have in the area.

pohly commented:
> If CSI and Device Plugins use different terms (plugin and driver) and DRA wants to use either of them it's ok. If DRA wants to use something different, I'm not sure we want to add to the mess we already have in the area.

The terminology used in this PR is consistent with CSI. A "CSI driver" and "DRA driver" are more than just a kubelet plugin.

klueska commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -23,70 +23,87 @@ import (
 	store map[string][]*Plugin
 }
 
-// draPlugins map keeps track of all registered DRA plugins on the node
-// and their corresponding sockets.
-var draPlugins = &pluginsStore{}
```
This is now embedded in the `Manager` type -- great -- I have never liked that this was a global variable.

klueska commented on pkg/kubelet/cm/dra/plugin/registration.go:
```diff
@@ -72,17 +74,16 @@ type RegistrationHandler struct {
 // If a kubeClient is provided, then it synchronizes ResourceSlices
 // with the resource information provided by plugins.
-func NewRegistrationHandler(kubeClient kubernetes.Interface, getNode func() (*v1.Node, error), wipingDelay time.Duration) *RegistrationHandler {
+func NewRegistrationHandler(draPlugins *Store, kubeClient kubernetes.Interface, getNode func() (*v1.Node, error), wipingDelay time.Duration) *RegistrationHandler {
```
it feels awkward to have this take an already initialized (but empty) `*Store` object which it then caches locally and populates. Does it make sense to have this return `(*RegistrationHandler, *Store)` so that the component that actually populates this object owns this object and the one that consumes it is the one that caches a reference to it.

pohly replied:
The way I see it, RegistrationHandler is one layer above the central store. It would be possible to use a store without a registration handler, or with some other way of registering endpoints. With that in mind, constructing the store separately seems consistent.

If this feels like overkill, then let's eliminate the RegistrationHandler. The store can implement `cache.PluginHandler` directly.

pohly replied:
I do not particularly care about RegistrationHandler being separate. I kept it because it was that way before and only clarified the layering.

klueska replied:
I like the idea of the store implementing cache.PluginHandler directly

klueska replied:
Well at this point, I'd just argue that maybe `store` isn't the right name for this abstraction, but rather something like `PluginManager`.

pohly replied:
Sounds okay to me. You can never have too many managers... okay, perhaps you can.

pohly replied:
I did the rename in a separate commit because then the file renaming is handled better in a diff because the files are similar enough: https://github.com/kubernetes/kubernetes/pull/132096/commits/dec98e8eea5569ad6f43f037032096fe2352d359

klueska reviewed (commented):
Left a few comments, but overall looks great. Thanks for the cleanup!

pohly committed:
DRA kubelet: rename Store + Plugin -> DRAPluginManager + DRAPlugin

After merging with RegistrationHandler, the store abstraction is more than a
dump map. It implements additional logic, so renaming it seems warranted.
To avoid confusion with other "plugin managers", the DRA prefix is used.
The plugin type gets updated accordingly.

This is done in a separate commit for ease of review. The "store" field is
kept because that really is just a dump lookup structure.

klueska commented on pkg/kubelet/cm/dra/plugin/dra_plugin_manager.go:
There are still some concerns with the approach taken here to introduce an overall `DRAPluginMananger` abstraction as a replacement for the separate `PluginStore` and `PluginRegistration` abstractions: https://github.com/kubernetes/kubernetes/pull/132058#discussion_r2128115743

However, I find merging them into a single entity both logical and elegant, since it now puts all Plugin management in a single place and removes the need for cross-component locking to ensure specific operations happen atomically. 

As such, I am inclined to approve this PR as-is.

We can always revisit this and separate them again in the future if we find a good reason to do so (e.g. testing becomes too convoluted).

klueska commented:
/lgtm
/approve
